### PR TITLE
Prevent upload same file with different template

### LINKED ISSUE #4076: Duplicate file uploads possible when coming from files section and files field
## Description

Following my [post on Nolt](https://kirby.nolt.io/388): 

When I upload a file from a files field that was already uploaded on that page before, the previously uploaded file gets replaced (or at least the creation date and time are updated), but existing metadata is attached to the new file.

This can lead to unwanted behavior. In my case, I uploaded a file from a files field that is set to use a specific template (using the query and uploads option). But that file was already added to the page using a files section—set to a different template.
The new (same) file was uploaded, but it was assigned the existing field values of the previous file including e. g. the template name, without any info or warning.

It works correctly if the file only has the same name, but not if it’s the exact same file.

**Expected behavior**  

Prevention of duplicate file uploads in that case.

## To reproduce

1. Create a blueprint having a files section and a files field set to use different templates:

```yaml
title: Default Page
sections:
  filesSection:
    type: files
    template: file-a
  fieldsSection:
    type: fields
    fields:
      filesField:
        type: files
        query: page.files.filter('template', 'file-b')
        uploads: file-b
```
2. Upload a file in the files section – it will use `file-a` as template as expected.
3. Upload that same file using the files field.

## Your setup
**Kirby Version**  

3.6.1.1

**Your system (please complete the following information)**
 - Device: MacBook Pro
 - OS: macOS 12.1
 - Browser: Google Chrome
 - Version: 96.0.4664.110

### PR DESCRIPTION
**Pull request by distantnative**
### Fixes
- When uploading the same file again (same filename, same content) an error is thrown when the assigned template doesn't match with the existing file
https://github.com/getkirby/kirby/issues/4076

## Ready?

- [x] Unit tests for fixed bug/feature
- [x] In-code documentation (wherever needed)
- [x] Tests and checks all pass

### For review team

- [x] Add changes to release notes draft in Notion
- [ ] Add to [website docs release checklist](https://github.com/getkirby/getkirby.com/pulls) (if needed)

### CODE REVIEW

File: src/Cms/FileRules.php [hunk 1]
```diff
@@ -80,8 +80,21 @@ public static function changeSort(File $file, int $sort): bool
...
     public static function create(File $file, BaseFile $upload): bool
     {
+        // first we want to ensure that we are not trying to
+        // create a duplicate file. So if a file
+        // with the same name already exists
         if ($file->exists() === true) {
-            if ($file->sha1() !== $upload->sha1()) {
+            // $file will be based on the props of the new file,
+            // so we need to get the props of the existing file
+            $existing = kirby()->file($file->id());
```
Comment (distantnative) [inline src/Cms/FileRules.php]: If somebody knows a better way for this, would be great

Comment (afbora) [reply src/Cms/FileRules.php]: Accessing the file via the parent is faster (microseconds 😅) than accessing from the Kirby instance. I'm not sure which one more stable.

````php
$existing = $file->parent()->file($file->filename());
````

File: src/Cms/FileRules.php [hunk 2]
```diff
@@ -80,8 +80,21 @@ public static function changeSort(File $file, int $sort): bool
...
+
+            // if the new upload is the exact same file
+            // and uses the same template, we can continue,
+            // otherwise throw an eroror for duplicate file
```
Comment (afbora) [inline src/Cms/FileRules.php]:
```suggestion
            // otherwise throw an error for duplicate file
```

**Review by afbora — DISMISSED** [review]

File: tests/Cms/Files/FileRulesTest.php [hunk 3]
```diff
@@ -113,6 +122,69 @@ public function testCreateExistingFile()
...
+            'parent' => $page,
+            'content' => [
+                'template' => 'cover'
+            ]
+        ]);
+
+        $upload = new BaseFile($testImage);
+        FileRules::create($newFile, $upload);
+
+        // same filename, same template and different file (sha1)
```
Comment (distantnative) [inline tests/Cms/Files/FileRulesTest.php]: I think this needs to be in a separate method as the method ends when the first exception is thrown.

Comment (afbora) [reply tests/Cms/Files/FileRulesTest.php]: So do we need:

- 3 unit tests? For `same file`, `same file and different template` and `same filename, same template and different file (sha1)`? 

- 2 unit tests? For `true` and `exceptions`.

Comment (distantnative) [reply tests/Cms/Files/FileRulesTest.php]: Maybe best to put them in three methods, yes

**Review by afbora — APPROVED** [review]

**Pull request merged**